llvm/lib/Target/AMDGPU/AMDGPUHotBlockRematerialize.cpp

```diff
@@ -0,0 +1,1511 @@
+//===- AMDGPUHotBlockRematerialize.cpp - AMDGPU Hot BlockRematerialize ----===//
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
+// License. See LICENSE.TXT for details.
+//
+//===----------------------------------------------------------------------===//
+//
+/// \file
+/// \brief AMDGPU hot block Rematerialize
```

Comment:
Can we do this in the IR? The complexity is about 100x higher to do this on MachineIR while dealing with LiveIntervals

Reply:
Agreed, it would make certain things much easier. It's been tried, but measuring liveness was an issue. It was also difficult to get an accurate picture of the register pressure, partly because the distinction between uniform vector operations and scalar operations was unclear.